pennylane_lightning/src/tests/Test_AVXSingleQubitGateHelpers.cpp

```diff
@@ -0,0 +1,344 @@
+#include "cpu_kernels/avx_common/SingleQubitGateHelper.hpp"
+
+#include <catch2/catch.hpp>
+
+using namespace Pennylane::Gates::AVXCommon;
```

Comment:
Not relevant now, but at a later point in time we can look into the Intel simulator at some point to validate we hit the YMM/ZMM registers with these functions.

Reply:
This is a generated assembly for the RX gate (with AVX512). It seems like it uses AVX registers extensively (still, I cannot tell if this is the most efficient code).

```
0000000000000000 <void Pennylane::Gates::AVXCommon::ApplyRX<double, 8ul>::applyInternal<0ul, double>(std::complex<double>*, unsigned long, bool, double)>:
   0:	f3 0f 1e fa          	endbr64 
   4:	4c 8d 54 24 08       	lea    0x8(%rsp),%r10
   9:	48 83 e4 c0          	and    $0xffffffffffffffc0,%rsp
   d:	41 ff 72 f8          	push   -0x8(%r10)
  11:	55                   	push   %rbp
  12:	48 89 e5             	mov    %rsp,%rbp
  15:	41 52                	push   %r10
  17:	48 81 ec e8 01 00 00 	sub    $0x1e8,%rsp
  1e:	48 89 bd 38 fe ff ff 	mov    %rdi,-0x1c8(%rbp)
  25:	48 89 b5 30 fe ff ff 	mov    %rsi,-0x1d0(%rbp)
  2c:	89 d0                	mov    %edx,%eax
  2e:	c5 fb 11 85 20 fe ff 	vmovsd %xmm0,-0x1e0(%rbp)
  35:	ff 
  36:	88 85 2c fe ff ff    	mov    %al,-0x1d4(%rbp)
  3c:	64 48 8b 04 25 28 00 	mov    %fs:0x28,%rax
  43:	00 00 
  45:	48 89 45 c8          	mov    %rax,-0x38(%rbp)
  49:	31 c0                	xor    %eax,%eax
  4b:	80 bd 2c fe ff ff 00 	cmpb   $0x0,-0x1d4(%rbp)
  52:	74 2b                	je     7f <void Pennylane::Gates::AVXCommon::ApplyRX<double, 8ul>::applyInternal<0ul, double>(std::complex<double>*, unsigned long, bool, double)+0x7f>
  54:	c5 fb 10 85 20 fe ff 	vmovsd -0x1e0(%rbp),%xmm0
  5b:	ff 
  5c:	c5 fb 10 0d 00 00 00 	vmovsd 0x0(%rip),%xmm1        # 64 <void Pennylane::Gates::AVXCommon::ApplyRX<double, 8ul>::applyInternal<0ul, double>(std::complex<double>*, unsigned long, bool, double)+0x64>
  63:	00 
  64:	c5 fb 5e e1          	vdivsd %xmm1,%xmm0,%xmm4
  68:	c4 e1 f9 7e e0       	vmovq  %xmm4,%rax
  6d:	c4 e1 f9 6e c0       	vmovq  %rax,%xmm0
  72:	ff 15 00 00 00 00    	call   *0x0(%rip)        # 78 <void Pennylane::Gates::AVXCommon::ApplyRX<double, 8ul>::applyInternal<0ul, double>(std::complex<double>*, unsigned long, bool, double)+0x78>
  78:	c4 e1 f9 7e c0       	vmovq  %xmm0,%rax
  7d:	eb 3f                	jmp    be <void Pennylane::Gates::AVXCommon::ApplyRX<double, 8ul>::applyInternal<0ul, double>(std::complex<double>*, unsigned long, bool, double)+0xbe>
  7f:	c5 fb 10 85 20 fe ff 	vmovsd -0x1e0(%rbp),%xmm0
  86:	ff 
  87:	c5 fb 10 0d 00 00 00 	vmovsd 0x0(%rip),%xmm1        # 8f <void Pennylane::Gates::AVXCommon::ApplyRX<double, 8ul>::applyInternal<0ul, double>(std::complex<double>*, unsigned long, bool, double)+0x8f>
  8e:	00 
  8f:	c5 fb 5e e9          	vdivsd %xmm1,%xmm0,%xmm5
  93:	c4 e1 f9 7e e8       	vmovq  %xmm5,%rax
  98:	c4 e1 f9 6e c0       	vmovq  %rax,%xmm0
  9d:	ff 15 00 00 00 00    	call   *0x0(%rip)        # a3 <void Pennylane::Gates::AVXCommon::ApplyRX<double, 8ul>::applyInternal<0ul, double>(std::complex<double>*, unsigned long, bool, double)+0xa3>
  a3:	c4 e1 f9 7e c0       	vmovq  %xmm0,%rax
  a8:	c5 fa 7e 05 00 00 00 	vmovq  0x0(%rip),%xmm0        # b0 <void Pennylane::Gates::AVXCommon::ApplyRX<double, 8ul>::applyInternal<0ul, double>(std::complex<double>*, unsigned long, bool, double)+0xb0>
  af:	00 
  b0:	c4 e1 f9 6e d8       	vmovq  %rax,%xmm3
  b5:	c5 e1 57 d0          	vxorpd %xmm0,%xmm3,%xmm2
  b9:	c4 e1 f9 7e d0       	vmovq  %xmm2,%rax
  be:	48 89 85 48 fe ff ff 	mov    %rax,-0x1b8(%rbp)
  c5:	c5 fb 10 85 20 fe ff 	vmovsd -0x1e0(%rbp),%xmm0
  cc:	ff 
  cd:	c5 fb 10 0d 00 00 00 	vmovsd 0x0(%rip),%xmm1        # d5 <void Pennylane::Gates::AVXCommon::ApplyRX<double, 8ul>::applyInternal<0ul, double>(std::complex<double>*, unsigned long, bool, double)+0xd5>
  d4:	00 
  d5:	c5 fb 5e f1          	vdivsd %xmm1,%xmm0,%xmm6
  d9:	c4 e1 f9 7e f0       	vmovq  %xmm6,%rax
  de:	c4 e1 f9 6e c0       	vmovq  %rax,%xmm0
  e3:	ff 15 00 00 00 00    	call   *0x0(%rip)        # e9 <void Pennylane::Gates::AVXCommon::ApplyRX<double, 8ul>::applyInternal<0ul, double>(std::complex<double>*, unsigned long, bool, double)+0xe9>
  e9:	c4 e1 f9 7e c0       	vmovq  %xmm0,%rax
  ee:	c4 e1 f9 6e c0       	vmovq  %rax,%xmm0
  f3:	ff 15 00 00 00 00    	call   *0x0(%rip)        # f9 <void Pennylane::Gates::AVXCommon::ApplyRX<double, 8ul>::applyInternal<0ul, double>(std::complex<double>*, unsigned long, bool, double)+0xf9>
  f9:	62 f1 fd 48 29 85 90 	vmovapd %zmm0,-0x170(%rbp)
 100:	fe ff ff 
 103:	48 8b 85 48 fe ff ff 	mov    -0x1b8(%rbp),%rax
 10a:	c4 e1 f9 6e c0       	vmovq  %rax,%xmm0
 10f:	ff 15 00 00 00 00    	call   *0x0(%rip)        # 115 <void Pennylane::Gates::AVXCommon::ApplyRX<double, 8ul>::applyInternal<0ul, double>(std::complex<double>*, unsigned long, bool, double)+0x115>
 115:	62 f1 fd 48 29 85 d0 	vmovapd %zmm0,-0x130(%rbp)
 11c:	fe ff ff 
 11f:	48 c7 85 40 fe ff ff 	movq   $0x0,-0x1c0(%rbp)
 126:	00 00 00 00 
 12a:	e9 aa 00 00 00       	jmp    1d9 <void Pennylane::Gates::AVXCommon::ApplyRX<double, 8ul>::applyInternal<0ul, double>(std::complex<double>*, unsigned long, bool, double)+0x1d9>
 12f:	48 8b 85 40 fe ff ff 	mov    -0x1c0(%rbp),%rax
 136:	48 c1 e0 04          	shl    $0x4,%rax
 13a:	48 89 c2             	mov    %rax,%rdx
 13d:	48 8b 85 38 fe ff ff 	mov    -0x1c8(%rbp),%rax
 144:	48 01 d0             	add    %rdx,%rax
 147:	48 89 c7             	mov    %rax,%rdi
 14a:	ff 15 00 00 00 00    	call   *0x0(%rip)        # 150 <void Pennylane::Gates::AVXCommon::ApplyRX<double, 8ul>::applyInternal<0ul, double>(std::complex<double>*, unsigned long, bool, double)+0x150>
 150:	62 f1 fd 48 29 85 50 	vmovapd %zmm0,-0x1b0(%rbp)
 157:	fe ff ff 
 15a:	62 f1 fd 48 28 85 50 	vmovapd -0x1b0(%rbp),%zmm0
 161:	fe ff ff 
 164:	62 f1 fd 48 59 85 90 	vmulpd -0x170(%rbp),%zmm0,%zmm0
 16b:	fe ff ff 
 16e:	62 f1 fd 48 29 85 10 	vmovapd %zmm0,-0xf0(%rbp)
 175:	ff ff ff 
 178:	48 8d 85 50 fe ff ff 	lea    -0x1b0(%rbp),%rax
 17f:	48 89 c7             	mov    %rax,%rdi
 182:	ff 15 00 00 00 00    	call   *0x0(%rip)        # 188 <void Pennylane::Gates::AVXCommon::ApplyRX<double, 8ul>::applyInternal<0ul, double>(std::complex<double>*, unsigned long, bool, double)+0x188>
 188:	62 f1 fd 48 59 85 d0 	vmulpd -0x130(%rbp),%zmm0,%zmm0
 18f:	fe ff ff 
 192:	62 f1 fd 48 29 85 50 	vmovapd %zmm0,-0xb0(%rbp)
 199:	ff ff ff 
 19c:	62 f1 fd 48 28 85 10 	vmovapd -0xf0(%rbp),%zmm0
 1a3:	ff ff ff 
 1a6:	62 f1 fd 48 58 85 50 	vaddpd -0xb0(%rbp),%zmm0,%zmm0
 1ad:	ff ff ff 
 1b0:	48 8b 85 40 fe ff ff 	mov    -0x1c0(%rbp),%rax
 1b7:	48 c1 e0 04          	shl    $0x4,%rax
 1bb:	48 89 c2             	mov    %rax,%rdx
 1be:	48 8b 85 38 fe ff ff 	mov    -0x1c8(%rbp),%rax
 1c5:	48 01 d0             	add    %rdx,%rax
 1c8:	48 89 c7             	mov    %rax,%rdi
 1cb:	ff 15 00 00 00 00    	call   *0x0(%rip)        # 1d1 <void Pennylane::Gates::AVXCommon::ApplyRX<double, 8ul>::applyInternal<0ul, double>(std::complex<double>*, unsigned long, bool, double)+0x1d1>
 1d1:	48 83 85 40 fe ff ff 	addq   $0x4,-0x1c0(%rbp)
 1d8:	04 
 1d9:	48 8b 85 30 fe ff ff 	mov    -0x1d0(%rbp),%rax
 1e0:	ba 01 00 00 00       	mov    $0x1,%edx
 1e5:	89 c1                	mov    %eax,%ecx
 1e7:	d3 e2                	shl    %cl,%edx
 1e9:	89 d0                	mov    %edx,%eax
 1eb:	89 c0                	mov    %eax,%eax
 1ed:	48 39 85 40 fe ff ff 	cmp    %rax,-0x1c0(%rbp)
 1f4:	0f 82 35 ff ff ff    	jb     12f <void Pennylane::Gates::AVXCommon::ApplyRX<double, 8ul>::applyInternal<0ul, double>(std::complex<double>*, unsigned long, bool, double)+0x12f>
 1fa:	90                   	nop
 1fb:	48 8b 45 c8          	mov    -0x38(%rbp),%rax
 1ff:	64 48 2b 04 25 28 00 	sub    %fs:0x28,%rax
 206:	00 00 
 208:	74 06                	je     210 <void Pennylane::Gates::AVXCommon::ApplyRX<double, 8ul>::applyInternal<0ul, double>(std::complex<double>*, unsigned long, bool, double)+0x210>
 20a:	ff 15 00 00 00 00    	call   *0x0(%rip)        # 210 <void Pennylane::Gates::AVXCommon::ApplyRX<double, 8ul>::applyInternal<0ul, double>(std::complex<double>*, unsigned long, bool, double)+0x210>
 210:	4c 8b 55 f8          	mov    -0x8(%rbp),%r10
 214:	c9                   	leave  
 215:	49 8d 62 f8          	lea    -0x8(%r10),%rsp
 219:	c3                   	ret    

```